### PR TITLE
275: Update user's password

### PR DESCRIPTION
**Pull request by dulowski-marek**
## Summary of change
This exposes an endpoint: `PUT /recipe/user/credentials` accepting the following body:
```ts
{
    "userId": string, // required
    "email": string, 
    "password": string
}
```

## Related issues
- https://github.com/supertokens/supertokens-core/issues/275

## Test Plan
tbd

## Documentation changes
- CDI changes

## Checklist for important updates
- [x] Changelog has been updated
- [ ] ~`coreDriverInterfaceSupported.json` file has been updated (if needed)~
- [ ] ~Changes to the version if needed~
   - ~In `build.gradle`~
- [x] Had installed and ran the pre-commit hook
- [ ] ~If there are new dependencies that have been added in `build.gradle`, please make sure to add them in `implementationDependencies.json`.~
- [x] Issue this PR against the latest non released version branch.
   - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
   - If no such branch exists, then create one from the latest released branch.

## Remaining TODOs for this PR
- [x] Swagger CDI spec
- [x] API tests
- [x] Logic tests

### CODE REVIEW

**Comment by rishabhpoddar** [comment]
- Please start by making changes to the CDI spec.
- This API is supposed to allow changing of a user's password or email by simply editing those fields in the db. No need to create any additional token etc.. So an APi like:
   ```
   /recipe/user PUT
   {
      userId: string,
      newPassword?: string,
      email?: string
   }
   ```
   "?" means they are optional. If the user has provided both those fields, both of them are changed. And the response can be:
   ```
   {
      status: "OK" | "EMAIL_ALREADY_EXISTS_ERROR"
   }
   ```
   The password validator would need to run in the backend SDK inside the recipe function that is used to call this API.
- Due to this large difference between what was expected and what is implemented, you should close this PR and start this from scratch.

**Comment by dulowski-marek** [comment]
Why do you think such API would be better than 2 endpoints allowing to separately change user's credentials? How is that different from password reset?
How do you want to approach change password request validation, if you can change the email in the same request? 

There are reasons why I chose to do two separate requests and token generation for password change. I don't know yours for the API design that you proposed at this moment, so please explain.

**Comment by rishabhpoddar** [comment]
> Why do you think such API would be better than 2 endpoints allowing to separately change user's credentials?

Because this is not a password reset flow.. it's simply one of the CRUD operations we provide to the backend against the objects of this recipe.


> How do you want to approach change password request validation, if you can change the email in the same request?

What is password request validation?

> There are reasons why I chose to do two separate requests and token generation for password change. I don't know yours for the API design that you proposed at this moment, so please explain.

My logic is simply to provide CRUD operation on the objects of this recipe. That's all. The actual password reset flow is different. Happy to discuss on a call.

**Comment by dulowski-marek** [comment]
Then this PR provides simple CRUD without session removal logic.

File: CHANGELOG.md [hunk 1]
```diff
@@ -16,7 +16,8 @@ https://github.com/supertokens/supertokens-core/issues/255
 ### Added
 - Add `GET /recipe/users/by-email?email=john@example.com` endpoint for ThirdParty recipe to fetch all users with given
  email
-
+- Add dedicated `POST /recipe/user/password/change/token` and `POST /recipe/user/password/change` endpoints allowing to
```
Comment (rishabhpoddar) [inline CHANGELOG.md:19]: The changelog message is wrong.

File: src/main/java/io/supertokens/webserver/api/emailpassword/UpdateEmailOrPasswordAPI.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import java.io.IOException;
+
+public class UpdateEmailOrPasswordAPI extends WebserverAPI {
+    public UpdateEmailOrPasswordAPI(Main main) {
+        super(main, RECIPE_ID.EMAIL_PASSWORD.toString());
+    }
+
+    @Override
+    public String getPath() {
+        return "/recipe/user/credentials";
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/webserver/api/emailpassword/UpdateEmailOrPasswordAPI.java:40]: Should change to `/recipe/user`

File: src/main/java/io/supertokens/webserver/api/emailpassword/UpdateEmailOrPasswordAPI.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    @Override
+    protected void doPut(HttpServletRequest req, HttpServletResponse resp) throws IOException, ServletException {
+        JsonObject input = InputParser.parseJsonObjectOrThrowError(req);
+        String userId = InputParser.parseStringOrThrowError(input, "method", false);
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/webserver/api/emailpassword/UpdateEmailOrPasswordAPI.java:46]: - Mistake - no "method", should be "userId"?
- Not tested API properly.

File: src/test/java/io/supertokens/test/emailpassword/UpdateUsersEmailAndPasswordTest.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
+/*
```
Comment (rishabhpoddar) [inline src/test/java/io/supertokens/test/emailpassword/UpdateUsersEmailAndPasswordTest.java:1]: - Need to add API test as well
- Add a test to update info of a user if they don't exist - should return status: "UNKNOWN_USER_ID"

File: src/main/java/io/supertokens/webserver/api/emailpassword/UpdateEmailOrPasswordAPI.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        if (email == null && password == null) {
+            throw new ServletException(new BadRequestException("You have to provide either email or password."));
+        }
+
+        try {
+            EmailPassword.updateUsersEmailOrPassword(main, userId, email, password);
+
+            JsonObject result = new JsonObject();
+            result.addProperty("status", "OK");
+            super.sendJsonResponse(200, result, resp);
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/webserver/api/emailpassword/UpdateEmailOrPasswordAPI.java:61]: The API output should be 
```
{
   status: "OK" | "UNKNOWN_USER_ID_ERROR"
}
```

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/emailpassword/EmailPassword.java [hunk 6]
```diff
@@ -188,6 +188,27 @@ public static void resetPassword(Main main, String token, String password)
         }
     }
 
+    public static void updateUsersEmailOrPassword(Main main, @Nonnull String userId, @Nullable String email, @Nullable String password)
+            throws StorageQueryException, StorageTransactionLogicException {
+        EmailPasswordSQLStorage storage = StorageLayer.getEmailPasswordStorage(main);
+
+        storage.startTransaction(transaction -> {
+            if (email != null) {
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/emailpassword/EmailPassword.java:196]: - You will first want to take a lock on all this userId's row by using a select for update query (which should already exist in the plugin interface).
- Then if the userId doesn't exist, you would return `"UNKNOWN_USER_ID_ERROR"`, else continue.

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

**Review by rishabhpoddar — APPROVED** [review]

**Pull request merged**